Update installation chapter (#62)
esp-rs/book: src/installation/index.md

# Setting Up a Development Environment

With an understanding of the ecosystem surrounding Rust on Espressif chips, we can move on to actual development. If you are not aware of the two possible development approaches or do not understand the differences between writing `std` and `no_std` applications, please first read the [Ecosystem Overview] chapter.

Let's take a moment to discuss the Rust support for the different architectures of the Espressif chips and the different approaches in more detail. At this moment, Espressif SoCs are based on two different architectures: `RISC-V` and `Xtensa`. The support for those two architectures in the Rust programming language is very different.

[Ecosystem Overview]: ../overview/index.md

## RISC-V targets

The `RISC-V` architecture has support in the mainline Rust compiler so, the setup is relatively simple. There are two ways of proceeding with the installation:

> - Using [`espup`, a tool that will be covered later]
> - Using the official Rust tools

Reviewer: ```suggestion
- Using the official Rust tools
- Using [`espup`, a tool that will be covered later]
```

I think we should encourage using official tools whenever possible, and as such it should be mentioned first IMO

If you only want to use `RISC-V` targets, you can use the official Rust tools, for this approach we need [`rustup`] installed, and a [Rust nightly toolchain] with the `rust-src` [component]. We can install a nightly toolchain with the `rust-src` component via:

```bash
rustup toolchain install nightly --component rust-src
```

> This are the two recommended targets for most Espressif `RISC-V` chips:

Reviewer: ```suggestion
These are the two recommended targets for most Espressif `RISC-V` chips:
```

> - For bare-metal (`no_std`) applications: `riscv32imc-unknown-none-elf`

Reviewer 2: I personally still feel like we should be installing the imac target by default, but we should discuss this in one of our roadmap meetings.

Author: Updated it to `riscv32imac-unknown-none-elf`

Reviewer 2: Really really sorry, but can you change it back :laughing:, I just did a couple of PRs to make imc the default for C3 and C2, see https://github.com/esp-rs/esp-template/pull/51 and https://github.com/esp-rs/esp-hal/pull/370

- For applications that require `std`: `riscv32imc-esp-espidf`

> #### A note in RISC-V `no_std` Rust targets.
>
> There are 3 different flavors of RISC-V 32 target in Rust:

> > - `riscv32i-unknown-none-elf`

Reviewer: Not sure we really need to mention this, it's implicitly supported and I'm not sure anybody actually uses it for anything (or would want to).

> - `riscv32imc-unknown-none-elf`
> - `riscv32imac-unknown-none-elf`
>
> Where:
> - `32i`: Base Integer Instruction Set for 32 bits
> - `m`: Standard Extension for Integer Multiplication and Division
> - `a`: Standard Extension for Atomic Instructions
>   - Only available in ESP32-C6 at the moment

> > - `c`: Standard Extension for Compressed Instructions

Reviewer: I actually think this whole section is kind of unneccessary, and we should probably just have a small blurb and then link out to some page which covers RISC-V extensions like this one:  
https://en.wikichip.org/wiki/risc-v/standard_extensions

The bare-metal targets can be installed by running:

```bash
rustup target add riscv32imc-unknown-none-elf
```

> For `std` applications, the `riscv32imc-esp-espidf` target is currently [Tier 3] and does not have prebuilt objects distributed through `rustup`, therefore, it does not need to be installed as the `no_std` targets. Furthermore, `std` projects, also require:

Reviewer: ```suggestion
For `std` applications, the `riscv32imc-esp-espidf` target is currently [Tier 3] and does not have pre-built objects distributed through `rustup`, therefore, it does not need to be installed as the `no_std` targets. Furthermore, `std` projects, also require:
```

[release binaries]: https://github.com/esp-rs/espup/releases
[cargo-binstall]: https://github.com/cargo-bins/cargo-binstall

### Other installation methods

- Using [esp-rs/rust-build] installation scripts. This was the recommended way in the past, but now the installation scripts are feature frozen and all new features will only be included in `espup`. See the repository README for instructions.
- Building the Rust compiler with `Xtensa` support from source. This process is computationally expensive and can take one or more hours to complete depending on your system. It is not recommended unless there is a major reason to go for this approach. See instructions in the [Installing from Source section of the esp-rs/rust repository].

[esp-rs/rust-build]: https://github.com/esp-rs/rust-build
[Installing from Source section of the esp-rs/rust repository]: https://github.com/esp-rs/rust#installing-from-source

## Rust with `std` runtime

Regardless of the target architecture, if you want to build a project using the [`std` approach], you will also need:
- [ESP-IDF]: Espressif IoT Development Framework as it's used as our hosted environment.

>   - See [ESP-IDF installation note] for details on how ESP-IDF can be installed.

Reviewer: Maybe this can be expanded on, you don't actually need to clone  ESP-IDF at all do you? I think that `embuild` clones it into the `.embuild` directory?

## RISC-V targets

The `RISC-V` architecture has support in the mainline Rust compiler so, the setup is relatively simple. There are two ways of proceeding with the installation:
- Using [`espup`, a tool that will be covered later]
- Using the official Rust tools

If you only want to use `RISC-V` targets, you can use the official Rust tools, for this approach we need [`rustup`] installed, and a [Rust nightly toolchain] with the `rust-src` [component]. We can install a nightly toolchain with the `rust-src` component via:

```bash
rustup toolchain install nightly --component rust-src
```

This are the two recommended targets for most Espressif `RISC-V` chips:

- For bare-metal (`no_std`) applications: `riscv32imc-unknown-none-elf`
- For applications that require `std`: `riscv32imc-esp-espidf`

> #### A note in RISC-V `no_std` Rust targets.
>
> There are 3 different flavors of RISC-V 32 target in Rust:
> - `riscv32i-unknown-none-elf`
> - `riscv32imc-unknown-none-elf`
> - `riscv32imac-unknown-none-elf`
>
> Where:
> - `32i`: Base Integer Instruction Set for 32 bits
> - `m`: Standard Extension for Integer Multiplication and Division
> - `a`: Standard Extension for Atomic Instructions

> >   - Only available in ESP32-S2 at the moment

Reviewer 3: guess ESP32-S2 doesn't belong here

Author: Upps, atomics are supported on ESP32 and ESP32-S3, right?

Good catch!

Reviewer 3: Yeah, but since this is the chapter about RISCV no need to mention them here
C6 supports the A extension AFAIK
